File: csharp/lesson 9/LowestCommonAncestor.cs
// https://leetcode.com/problems/lowest-common-ancestor-of-a-binary-tree

public class Solution
{
    public TreeNode LowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) 
    {
        if (root == null || root == p || root == q) return root;

        var left = LowestCommonAncestor(root.left, p, q);
        var right = LowestCommonAncestor(root.right, p, q);

        if (left != null && right != null) return root;
        return left ?? right;
    }
}

File: kotlin/lesson 9/LowestCommonAncestor.kt
// https://leetcode.com/problems/lowest-common-ancestor-of-a-binary-tree

class Solution {
    fun lowestCommonAncestor(root: TreeNode?, p: TreeNode?, q: TreeNode?): TreeNode? {
        if (root == null || root == p || root == q) return root

        val left = lowestCommonAncestor(root.left, p, q)
        val right = lowestCommonAncestor(root.right, p, q)

        return if (left != null && right != null) {
            root
        } else if (left != null) {
            left
        } else {
            right
        }
    }
}
